In [3]:
import numpy as np
import math
import sympy
from sympy.abc import x,y
from sympy import *
init_printing( use_latex='mathjax' )
f = Function("f")

$$I = \int_{\pi/4}^{5\pi/4} |sin(sin(x))|dx$$

Формула для составного метода Эйлера-Маклорена равна:

$$\int_{a}^{b} f(x)dx = \frac{h}{2} (f_0 + 2f_1 + ... + 2f_{n-1} + f_n) + \frac{h^2}{2}(f'_0 - f'_n)$$

Функция не является дифференцируемой на отрезке, а значит что оценку нужно будет получить при помощи процесса Эйткена

In [4]:
f = abs(sin(sin(x)))
f

│sin(sin(x))│

Поскольку в точке $\frac{\pi}{4}$ и ее окрестности наша функция положительна, то ее производная в $\frac{\pi}{4}$ равна производной $sin(sin(x))$, в точке $\frac{5\pi}{4}$ иная ситуация, там функция и ее окрестность строго меньше нуля, тогда производная $f$ в $\frac{5\pi}{4}$ равна производной $-sin(sin(x))$


Однако, т.к.

In [5]:
diff(sin(sin(x)), x)

cos(x)⋅cos(sin(x))

То, нехитрыми тригонометрическими действиями легко выяснить, что $cos(\pi/4) \cdot cos(sin(\pi/4)) = -cos(5\pi/4) \cdot cos(sin(5\pi/4))$

Значит производные на концах равны и $\frac{h^2}{2}(f'_0 - f'_n) = 0$

In [6]:
def EM(N1):
    point=[0]
    value=[f.subs(x, point[0])]
    for i in range(0,N1):
        point.append(point[i]+np.pi/N1)
        value.append(f.subs(x,point[i+1]))
    I = value[0] + value[N1-1]
    for i in range(1, N1-1):
        I += 2*value[i]
    return I * (np.pi/N1)/2

Уточним все с помощью процесса Эйткина

Поскольку у нас 4 порядок точности, то формула будет иметь вид:

$$\begin{cases}
  I = I_h + c \cdot h^4\\
  I = I_{2h} + c \cdot (2h)^4\\
\end{cases}$$

Откуда 
$$I = \frac{16I_h - I_{2h}}{15}$$

$$c = \frac{I_h - I_{2h}}{15h^4}$$

In [7]:
N=10
h=np.pi/N
print(h)
c=(EM(2*N)-EM(N))/(15*h**4)
abs(c)

0.3141592653589793


0.327988911518753

Ошибка получается

$$E = h^4 \cdot 0.328 \rightarrow 10^{-4} = h^4 \cdot 0.328 \rightarrow h \leq 0.132$$

Откуда получим что $N = \frac{\pi}{h} = 23.8$, значит берем $N = 24$

Тогда наш интеграл по формуле составного метода Эйлера-Маклорена равен

In [8]:
EM(24)

1.77511137630677

А уточнив полученное значение интеграла при помощи процесса Эйткена

In [9]:
(16*EM(24) - EM(12))/15

1.77734976009020